File: udp_example/udp_server.py
import sys
import socket

BUF_SIZE = 1024

if __name__ == "__main__":

  if len(sys.argv) != 2:
    sys.stderr.write("usage: udp_server port\n")
    exit(1)

  try:
    port = int(sys.argv[1])
    assert port > 0 
  except:
    sys.stderr.write("error: invalid port\n")
    exit(1)

  sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
  sock.bind(("0.0.0.0", port))

  while True:
    data, addr = sock.recvfrom(BUF_SIZE)
    print(addr[0], "sent a message.")  

    sock.sendto(data, addr)

  sock.close()

File: udp_example/udp_server.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

int main(int argc, char * argv[]) {
  int BUF_SIZE = 1024;
  int port;
  int sockfd;
  struct sockaddr_in listen_addr;
  struct sockaddr_in client_addr;
  int sin_size, recv_size;
  char buffer[BUF_SIZE+1];

  if (argc != 2) { 
    fprintf(stderr,"usage: udp_server port\n");
    exit(1);
  }
  port = atoi(argv[1]);
  if (port <= 0) {
    fprintf(stderr,"error: invalid port\n");
    exit(1);
  }
  if ((sockfd = socket(AF_INET, SOCK_DGRAM, 0)) == -1) {
    perror("socket");
    exit(1);
  }

  listen_addr.sin_family = AF_INET;
  listen_addr.sin_port = htons(port);
  listen_addr.sin_addr.s_addr = INADDR_ANY;

  if (bind(sockfd, (struct sockaddr *)&listen_addr, sizeof(struct sockaddr)) == -1) {
    perror("bind");
    exit(1);
  }

  while(1) { 
    sin_size = sizeof(struct sockaddr);
    if ((recv_size = recvfrom(sockfd, buffer, BUF_SIZE, 0, (struct sockaddr *)&client_addr, &sin_size)) == 0) {
      close(sockfd);
      if (errno != 0) {
        perror("recvfrom");
      }
      break;
    }
    buffer[recv_size] = '\0';
    printf("Received message. Size: %d, message: %s", recv_size, buffer);
    if (sendto(sockfd, buffer, recv_size, 0, (struct sockaddr *)&client_addr, sizeof(struct sockaddr)) == -1) {
      perror("sendto");
      close(sockfd);
      break;
    }
  }
}